frame/staking/src/lib.rs

```diff
@@ -807,3 +807,75 @@ where
 		R::is_known_offence(offenders, time_slot)
 	}
 }
+
+/// Something that can dictate the maximum number of nominations per nominator.
+pub trait NominationQuota<Balance> {
+	/// The absolute maximum number that this trait can return.
+	// NOTE: this is useful for generating the compact solution type, i.e.
+	// `sp_npos_elections::generate_solution_type`.
+	const ABSOLUTE_MAXIMUM: u32;
+
+	/// Determine the number of nominations that an account with `balance` at stake is allowed to
+	/// have.
+	fn nomination_quota(balance: Balance) -> u32;
+}
+
+/// A nomination quota descriptor that allows `MAX` for all nominators.
+pub struct FixedNominationQuota<const MAX: u32>;
+impl<Balance, const MAX: u32> NominationQuota<Balance> for FixedNominationQuota<MAX> {
+	const ABSOLUTE_MAXIMUM: u32 = MAX;
+
+	fn nomination_quota(_: Balance) -> u32 {
+		MAX
+	}
+}
+
+use sp_runtime::{
+	traits::{One, UniqueSaturatedInto},
+	SaturatedConversion,
+};
+
+/// An implementation of a linear nomination distribution.
+///
+/// TODO: ideally we want this to be a const expression all the way through, no point in
+/// re-computing this on the fly. For now we stick to a runtime placeholder.
+///
+///  Essentially, draws a line between `(MinBalance, MinQuota)` and `(MaxBalance, MaxQuota)`, Also,
+/// values less than `MinBalance` have `MinQuota` and values higher than `MaxBalance` have
+/// `MaxQuota`.
+pub struct LinearNominationQuota<MinBalance, MaxBalance, const MAX_QUOTA: u32, const MIN_QUOTA: u32>(
```

Comment:
I won't test this further _yet_ because I am not sure if we are going to use it.